doc/dev/rust-upgrade.md

```diff
@@ -0,0 +1,344 @@
+    with Rust changes via `git blame`. You should incorporate all the changes
+    from the "*Update*" lines into your current patch, and fold the update
+    descriptions into the main patch description, so that the patch will be one
+    single commit this time.
+
+    After making the patch changes, check that the patch at least compiles by
+    running `mir-json-translate-libs`, run relevant `crux-mir` tests if it is
+    easy to do, and then commit the patch. Make sure to update the "last
+    applied" date, and **always make a single commit for each patch**, with both
+    the `Patches.md` edits and the actual Rust changes. If later on you realize
+    some part of the patch needs to be changed, use `git rebase` to edit the
+    commit. If it is hard to rebase for some reason, then add an "*Update*" line
+    with your new change, again making a single commit containing both the
+    "*Update*" line and the code changes.
+
+    After applying all the existing patches, there may still be some `crux-mir`
+    test failures. This may mean that new patches are needed. Add new patches to
+    the bottom of `Patches.md` following the existing format. For each new
+    patch, make sure there is a `crux-mir` test that tests the feature it's
+    patching for. If the need for the new patch was only incidentally revealed
+    by the failure of a `crux-mir` test case that was mainly testing a different
+    thing, it's best to make a new `crux-mir` test for the patched feature
+    specifically, so that even if in the future the patch becomes unnecessary
+    for the original test case to work, the patch is still tested.
+
+    Creating new patches, or re-doing existing patches, might require changes on
+    the `crucible-mir` side. Commit on your `crucible` branch as you make those
+    changes.
+
+    As you look at the `crux-mir` test outputs, some tests are marked as
+    expected to fail and will say something like "FAIL (expected)" in the test
+    output. The test runner only checks that they fail and not *how* they
+    failed, so inspect the XFAIL test outputs to make sure that the failure
+    matches the reason described and is not some unexpected error.
+
+    Finally, some `crux-mir` test failures under the `symb_eval` directory may
+    be due to benign things like changes in line numbers in error messages for
+    certain things in the standard library. In that case, simply update the
+    `.good` file with the new output and commit.
+
+    After all the `crux-mir` tests pass, it's worth checking that the
+    `crux-mir-comp` tests pass as well. If there were any `crucible`-side
+    changes, point `saw-script`'s `crucible` dependency to your `crucible`
+    branch (e.g. by temporarily modifying `cabal.project`), then build and run
+    the `crux-mir-comp` tests. Make and commit necessary changes in `mir-json`,
+    `crucible`, and/or `saw-script` for the `crux-mir-comp` tests to pass
+    (except the `cabal.project` changes).
```

Comment:
Instead of temporarily modifying `cabal.project`, one can also make a `cabal.project.local` file (which is ignored by `.git`) containing an [optional dependency](https://cabal.readthedocs.io/en/stable/cabal-project-description-file.html#cfg-field-optional-packages) that points to the `crucible` branch.

Reply:
I've changed it to suggest updating the submodule instead, along with the other changes to the submodule workflow later on. (Basically I was originally trying to avoid having a lot of submodule updating and commit amending in the process because I thought it would be confusing, but it's not a big deal in the grand scheme of things.)